### ISSUE TITLE
JDL integration into generator-jhipster.

### ISSUE THREAD

**Issue by mshima** [jhipster/generator-jhipster#12388]
##### **Overview of the feature request**

Some integration/cleanup/implementation suggestions for JDL.

- [ ] Remove write support from jdl.
Entity and application file write was moved from jdl to import-jdl command.
This was done due to having old (existing file) and new state (config returned from jdl).
The missing part is deployment.
- [ ] Stop creating a creationTimestamp.
There is too many duplications for this feature and the generator can handle a missing creationTimestamp config.
- [ ] Stop setting default configuration.
`jhipster --defaults` uses current jdl defaults, so it could be unified by stop setting defaults at jdl and passing defaults option to the generator.
- [ ] Implement annotations at application.
Annotations are supported at entities, fields and relationships.
The missing part is application, this could be useful for blueprints, currently there is no way to add a blueprint configuration to jdl.

##### **Related issues or PR**

-   [ ] Checking this box is mandatory (this is just to show you read everything)

**Comment by pascalgrimaud** [comment]
cc @MathieuAA too

**Comment by MathieuAA** [comment]
> Implement annotations at application.
Annotations are supported at entities, fields and relationships.
The missing part is application, this could be useful for blueprints, currently there is no way to add a blueprint configuration to jdl.

I don't like this solution. It's easier to allow custom options inside configurations.

**Comment by MathieuAA** [comment]
> Stop setting default configuration.
jhipster --defaults uses current jdl defaults, so it could be unified by stop setting defaults at jdl and passing defaults option to the generator.

This is broader than just the JDL, it also involves the JHipster base repo and the generator.

**Comment by mshima** [comment]
> > Stop setting default configuration.
> > jhipster --defaults uses current jdl defaults, so it could be unified by stop setting defaults at jdl and passing defaults option to the generator.
> 
> This is broader than just the JDL, it also involves the JHipster base repo and the generator.

I completely disagree on this one.
- As far as I remember there was no discussion with core developers about adding dependency on jhipster-base.
- generator-jhipster should be self contained, please don't add another jhipster-core now that jhipster-core was merged.
- jhipster-base should depend on generator-jhipster, not generator-jhipster depend on jhipster-base.
- jhipster supports prompts too, same defaults used by jdl should be used by prompts.

This way we will start creating more complexities on our development workflow.
Anyway, this is a suggestion, feel free to add needs-discussion to it.

**Comment by mshima** [comment]
> > Implement annotations at application.
> > Annotations are supported at entities, fields and relationships.
> > The missing part is application, this could be useful for blueprints, currently there is no way to add a blueprint configuration to jdl.
> 
> I don't like this solution. It's easier to allow custom options inside configurations.

Any implementation that adds blueprint custom options.

**Comment by mshima** [comment]
> > Implement annotations at application.
> > Annotations are supported at entities, fields and relationships.
> > The missing part is application, this could be useful for blueprints, currently there is no way to add a blueprint configuration to jdl.
> 
> I don't like this solution. It's easier to allow custom options inside configurations.

Custom options issue is open for more than 1 year https://github.com/jhipster/generator-jhipster/issues/9870.

**Comment by MathieuAA** [comment]
Marcelo, thanks for explaining your viewpoint on this. It's clear now there's a misunderstanding about JHipster base that should be cleared. I guess we @jhipster/developers should discuss it then.

JHipster Base is a side project that aims at being the sole place where options, values, fields, etc. should be. Basically like a lib that can be shared across projects (the generator, the JHipster ide, whatever). It should not contain any business logic at all. Not if-s, no callable function whatsoever, just a repository. Its goal is not the same as JCore's (RIP).

It cannot depend on any other project as it should be a provider of information, not a consumer of information.
It can either be in another repo, or inside the generator in a monorepo fashion (I favor the latter).

Is it clearer @mshima?

**Comment by mshima** [comment]
It should be a monorepo IMO.
I think we should migrate to npm 7 workspace already.

> Is it clearer @mshima?

Sure, but IMO that don't change the fact that if we set defaults at jdl parsing, it will be duplicated.

I completely agree we should create a kind of jhipster-base but it should be inside generator-jhispter repository.

It should unify prompt question and options:
https://github.com/jhipster/generator-jhipster/blob/9b0518ff80406fa424a79c1555048db0cf9b9ff0/generators/app/prompts.js#L48-L65
https://github.com/jhipster/generator-jhipster/blob/9b0518ff80406fa424a79c1555048db0cf9b9ff0/jdl/jhipster/application-types.js#L20-L25

It should unify option description:
https://github.com/jhipster/generator-jhipster/blob/9b0518ff80406fa424a79c1555048db0cf9b9ff0/generators/app/index.js#L52-L55

It should unify defaults:
https://github.com/jhipster/generator-jhipster/blob/9036bbf61c9959a9b8c30c92332d71c55b6d0b42/generators/generator-base.js#L2370-L2389
https://github.com/jhipster/generator-jhipster/blob/8a939555af2933247174b3c627d7409ac7327ee6/jdl/models/jdl-application-factory.js#L39-L60

And so one.

**Comment by colameo** [comment]
> Marcelo, thanks for explaining your viewpoint on this. It's clear now there's a misunderstanding about JHipster base that should be cleared. I guess we @jhipster/developers should discuss it then.
> 
> JHipster Base is a side project that aims at being the sole place where options, values, fields, etc. should be. Basically like a lib that can be shared across projects (the generator, the JHipster ide, whatever). It should not contain any business logic at all. Not if-s, no callable function whatsoever, just a repository. Its goal is not the same as JCore's (RIP).
> 
> It cannot depend on any other project as it should be a provider of information, not a consumer of information.
> It can either be in another repo, or inside the generator in a monorepo fashion (I favor the latter).
> 
> Is it clearer @mshima?
> 

All Jhipster base "models" are used in the JH IDE to validate, for showing tooltip messages and are also fully integrated in the core part of the JDL specification. Also to mention, that we generate MD files from those models for documentation purpose. 

So the idea that @MathieuAA and I had was to have one single source/place/repo for defining such shared models.

**Comment by pascalgrimaud** [comment]
I'd like to know what is the state of jhipster-base today ? Still in dev ? Already used ? By which project ?

**Comment by colameo** [comment]
> I'd like to know what is the state of jhipster-base today ? Still in dev ? Already used ? By which project ?

used since 3 releases in JH IDE.

**Comment by mshima** [comment]
> All Jhipster base "models" are used in the JH IDE to validate, for showing tooltip messages and are also fully integrated in the core part of the JDL specification. Also to mention, that we generate MD files from those models for documentation purpose.
> 
> So the idea that @MathieuAA and I had was to have one single source/place/repo for defining such shared models.

That's great, but I remember asking for an approach more friendly with the generator.

Consider https://github.com/jhipster/jhipster-base/blob/ac65b5e96addb525f42c6764863119ab21606d1d/src/JDLApplicationOptions.json#L1-L13:
```
[
  {
    "option": "applicationType",
    "baseType": "STRING",
    "parseType": "EnumLiteral",
    "values": [
      "monolith",
      "microservice",
      "gateway",
      "uaa"
    ],
    "description": "One of the listed values"
  },
```

Should be something like
```
'applicationType' : {
    "option": "applicationType",
    "description": "Application type to generate"
    "question": "Which type of application would you like to create?",
    "default": "monolith",
    "values": [
      {
          value: "monolith",
          description: "Monolithic application (recommended for simple projects)"
     }, ...
    ],
    "prompt": {
        "type: "list"
    },
    "option": {
        "type": "String"
    },
    "jdl": {
        "baseType": "STRING",
        "parseType": "EnumLiteral",
        "description": "One of the listed values"
    },
  },
```

Integrating prompts:
```
const def = require('JDLApplicationOptions.json'). applicationType;
this.prompt(
        {
            type: def.prompt.type,
            name: 'applicationType',
            message: def.question,
            choices: def.values.map(alternative => ...),
            default: def.default,
        }
);
```

Integrating options:
```
const def = require('JDLApplicationOptions.json'). applicationType;
this.option(
            type: def.option.type,
            desc: `${def.description}  (values:  ${default.values.map(alternative => alternative.value)}, default: ${default.default})`
);
```

**Comment by pascalgrimaud** [comment]
Ok then, we have a lack of communication because, I'm not sure all the core team is aware of this project. There is some development in main generator-jhipster which impacts this project, so PR should be done there too. Here some examples :

- for v7, Vue.js should not be used in blueprint : https://github.com/jhipster/jhipster-base/blob/master/src/JDLApplicationOptions.json#L19

- for v7, the correct value is `vue`, not `vuejs` : https://github.com/jhipster/jhipster-base/blob/master/src/JDLApplicationOptions.json#L322

- for v7: missing Cypress: https://github.com/jhipster/jhipster-base/blob/master/src/JDLApplicationOptions.json#L70-L72

- missing Neo4j:
    - https://github.com/jhipster/jhipster-base/blob/master/src/JDLApplicationOptions.json#L51-L61
    - https://github.com/jhipster/jhipster-base/blob/master/src/JDLApplicationOptions.json#L187-L195

- no more question about useSass, it should be deleted: https://github.com/jhipster/jhipster-base/blob/master/src/JDLApplicationOptions.json#L336

- missing new languages : https://github.com/jhipster/jhipster-base/blob/master/src/JDLLanguages.json

It means, for each PR with new features, new questions, etc, we should check these 2 projects too:
- jhipster-online
- jhipster-base

**Comment by mshima** [comment]
> It means, for each PR with new features, new questions, etc, we should check these 2 projects too:
> 
> * jhipster-online
> * jhipster-base

@pascalgrimaud I actually don't agree with this approach.
Our official blueprints are supported too, so if that's the case, the new feature at approval should be supported by our blueprints too.
I did the generator refactoring, I not considering I will have to do all blueprint-* update by myself. Of corse I will help as much as possible.

Once we have an options structure on our main generator. It should be easy to automatise jhipster-online definitions update, jhipster-base too.

**Comment by pascalgrimaud** [comment]
@mshima : we didn't understand each other :-) Let me try to explain again:
- we add Vue.js, as new official option
- we need to add this option at start.jhipster.tech -> see my PR https://github.com/jhipster/jhipster-online/pull/188
- with jhipster-base, I need to do a PR at https://github.com/jhipster/jhipster-base/pulls
- we need to update the documation too at https://github.com/jhipster/jhipster.github.io

Another example, for Cypress, no PR at jhipster-online at the moment. So the authors of the PR should not forget this part

Hope I'm clearer :-)

**Comment by mraible** [comment]
Any updates y'all? Can we close this in favor of a pull request or two?

**Comment by deepu105** [comment]
@MathieuAA since the repo is already merged to master can we close this? Is there more work to do? i believe we need to figure out a release process for this

**Comment by mshima** [comment]
Those are just suggested to have a better integration between jdl and the generators.
